Unix/http/sessionmap.c

```diff
@@ -0,0 +1,318 @@
+            // From strset.c
+            hash ^= ((hash << 5) + *p + (hash >> 2));
+            p++;
+        }
+    }
+    return hash;
+}
+
+static size_t _SessionBucket_GetHash(_In_ const HashBucket *b)
+{
+    const SessionBucket *bucket = (const SessionBucket*) b;
+    return bucket->hash;
+}
+
+static int _SessionBucket_Equal(_In_ const HashBucket* b1, _In_ const HashBucket* b2)
+{
+    const SessionBucket* bucket1 = (const SessionBucket*) b1;
+    const SessionBucket* bucket2 = (const SessionBucket*) b2;
+    return 
+    (
+        bucket1->hash == bucket2->hash
+        &&
+        strcmp(bucket1->sessionId, bucket2->sessionId) == 0
+    );
+}
+
+static void _SessionBucket_Release(HashBucket *b)
+{
+    if (b)
+    {
+        SessionBucket* bucket = (SessionBucket*) b;
+        if (bucket->cookie)
+        {
+            PAL_Free(bucket->cookie);           
+        }
+        PAL_Free(bucket);
+    }
+}
+
+static SessionBucket* _SessionBucket_New(SessionMap *self, _In_z_ const char* sessionId, size_t hash)
+{
+    // WriteLock expected.
+    SessionBucket* bucket = NULL;
+    if (sessionId) do
+    {
+        size_t keyLen = strlen(sessionId);
+        size_t size = sizeof(SessionBucket) + keyLen + 1;
+        bucket = PAL_Malloc(size);
+        if (NULL == bucket)
+        {
+            LOGE2((ZT("_SessionBucket_New - failed to allocate bucket: %s"), sessionId));
+            DEBUG_ASSERT(FALSE);
+            break;
+        }
+
+        char* clone = (char*) (((unsigned char*) bucket) + sizeof(SessionBucket));
+        strcpy(clone, sessionId);
+        bucket->sessionId = clone;
+        bucket->hash = hash;
+        bucket->cookie = NULL;
+        bucket->next = NULL;
+
+        if (HashMap_Insert (&self->map, (HashBucket*) bucket) != 0)
+        {
+            // Not expecting the bucket to already exist.
+            _SessionBucket_Release((HashBucket*) bucket);
+            bucket = NULL;
+            LOGE2((ZT("_SessionBucket_New - sessionId already exists: %s"), sessionId));
+            DEBUG_ASSERT(FALSE);
+            break;
+        }
+        LOGD2((ZT("SessionBucket_New - added new session id: %s"), sessionId));
+    } while (0);
+
+    return bucket;
+}
+
+static BOOL _SessionMap_IsValid(SessionMap* self)
+{
+    if (NULL == self)
+    {
+        return FALSE;
+    }
+
+    ReadWriteLock_AcquireWrite(&self->lock);
+
+    int result = 0;
+    if (SESSIONID_HASH_SEED != self->state)
+    {
+        const size_t numItems = 64;
+
+        result = HashMap_Init(&self->map, 
+            numItems, 
+            _SessionBucket_GetHash, 
+            _SessionBucket_Equal, 
+            _SessionBucket_Release);
+
+        if (0 == result)
+        {
+            LOGD2((ZT("_SessionMap_IsValid - SessionMap initialized.")));
+            self->state = SESSIONID_HASH_SEED;
+        }
+        else
+        {
+            LOGE2((ZT("_SessionMap_IsValid - Could not initialize the HashMap.")));
+            DEBUG_ASSERT(FALSE);
```

Comment:
asserting on failure is wrong. Return an error and logging is sufficient.

Reply:
No fatal errors, right. :)